lib/ansible/modules/storage/zfs/zpool.py

```diff
@@ -0,0 +1,454 @@
+
+    def create(self):
+        if self.module.check_mode:
+            self.changed = True
+            return
+        cmd = [self.zpool_cmd]
+        if self.add:
+            action = 'add'
+        elif self.sets:
+            action = 'set'
+        else:
+            action = 'create'
+        cmd.append(action)
+        if action == 'create':
+            if self.ashift and action == 'create':
+                ashift = '-o ashift=' + str(self.ashift)
+            else:
+                ashift = '-o ashift=0'
+            cmd.append(ashift)
+        cmd.append(self.autoreplace)
+        cmd.append(self.autoexpand)
+        cmd.append(self.name)
+        if action != 'set':
+            cmd.append(self.devices)
+            cmd.append(self.spare)
+            cmd.append(self.zil)
+            cmd.append(self.l2arc)
+        (rc, out, err) = self.module.run_command(' '.join(cmd))
+        if rc == 0:
+            self.changed = True
+        else:
+            self.module.fail_json(msg=err)
+
+    def destroy(self):
+        if self.module.check_mode:
+            self.changed = True
+            return
+        cmd = [self.zpool_cmd, 'destroy', self.name]
+        (rc, out, err) = self.module.run_command(' '.join(cmd))
+        if rc == 0:
+            self.changed = True
+        else:
+            self.module.fail_json(msg=err)
+
+
+def path_leaf(path):
+    head, tail = ntpath.split(path)
+    return tail or ntpath.basename(head)
+
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(type='str', required=True),
+            state=dict(type='str', required=True, choices=['absent', 'present']),
+            raid_level=dict(type='str', required=False, choices=['raid0', 'mirror', 'raidz', 'raidz1', 'raidz2', 'raidz3']),
+            vdev=dict(type='int', require=False),
+            devices=dict(type='list', default=None),
+            spare=dict(type='list', default=None),
+            add=dict(type='bool', default=False),
+            ashift=dict(type='int', default=0, choices=[0, 9, 10, 11, 12, 13, 14, 15, 16]),
+            sets=dict(type='bool', default=False),
+            autoreplace=dict(type='bool'),
+            autoexpand=dict(type='bool'),
+            zil=dict(type='str', default=None),
+            l2arc=dict(type='str', default=None),
+        ),
+        supports_check_mode=True,
+        required_together=[['raid_level', 'devices'], ['devices', 'vdev']],
+        mutually_exclusive=[['add', 'sets']]
+    )
+
+    name = module.params.get('name')
+    state = module.params.get('state')
+    add = module.params.get('add')
+    raid_level = module.params.get('raid_level')
+    devices = module.params.get('devices')
+    spare = module.params.get('spare')
+    vdev = module.params.get('vdev')
+    ashift = module.params.get('ashift')
+    sets = module.params.get('sets')
+    autoreplace = module.params.get('autoreplace')
+    autoexpand = module.params.get('autoexpand')
```

Comment:
Instead of autoreplace/autoexpand, you should consider adding properties as a general mechanism to pass in settings or create an option for every ZFS property.